File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode ReverseBetween(ListNode head, int left, int right) {

       if (left == right || head == null || head.next == null) return head;

        // Create a dummy node to handle edge cases where the reversal starts at the head
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prev = dummy;

        // Move `prev` to the node just before the `left` position
        for (int i = 1; i < left; i++) {
            prev = prev.next;
        }

        // `start` is the first node in the sublist to be reversed
        ListNode start = prev.next;
        ListNode nodeToReverse = start.next;

        // Reverse the sublist between `left` and `right`
        for (int i = 0; i < right - left; i++) {
            start.next = nodeToReverse.next;
            nodeToReverse.next = prev.next;
            prev.next = nodeToReverse;
            nodeToReverse = start.next;
        }

        return dummy.next;
}
}

File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.py
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution(object):
    def reverseLL(self, newHead, newTail, beforeHead, afterTail):
        dummyCurr = newHead
        beforeCurr = None
        # reverse nodes from newHead up to (and including) newTail
        while dummyCurr != afterTail:
            temp = dummyCurr.next
            dummyCurr.next = beforeCurr
            beforeCurr = dummyCurr
            dummyCurr = temp

        # reconnect
        beforeHead.next = newTail
        newHead.next = afterTail
        return

    def reverseBetween(self, head, left, right):
        if not head or left == right:
            return head

        sentinel = ListNode(0, head)
        beforeHead = sentinel

        # move beforeHead to node just BEFORE position `left`
        for _ in range(left - 1):
            beforeHead = beforeHead.next

        newHead = beforeHead.next  # node at position `left`

        # move to node at position `right`
        newTail = newHead
        for _ in range(right - left):
            newTail = newTail.next

        afterTail = newTail.next

        self.reverseLL(newHead, newTail, beforeHead, afterTail)
        return sentinel.next
